### PR TITLE
Mangling: enable new mangling for symbols

### PR DESCRIPTION
**Pull request by eeckstein**
Finally
rdar://problem/24085509

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please test and merge

**Comment by eeckstein** [comment]
@swift-ci Please clean test linux

File: lib/Basic/Mangler.cpp [hunk 1]
```diff
@@ -32,6 +32,7 @@ llvm::cl::opt<bool> NewManglingForTests(
 
 #ifndef USE_NEW_MANGLING
 
+#ifndef USE_NEW_MANGLING
```
Comment (JaSpa) [inline lib/Basic/Mangler.cpp]: Seems like this `#ifndef` is a duplication

Comment (eeckstein) [inline lib/Basic/Mangler.cpp]: Thanks, I'll remove it

**Comment by eeckstein** [comment]
This PR must be in sync with https://github.com/apple/swift-corelibs-foundation/pull/919

**Comment by eeckstein** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/919

@swift-ci Please test

**Comment by eeckstein** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/919

@swift-ci Please test

**Comment by eeckstein** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/919

@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please clean test linux

**Comment by eeckstein** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/919

@swift-ci Please clean test linux

**Comment by jrose-apple** [comment]
Triple-checking: does this break serialized archives that have old mangled names in them, which will be looked up at runtime on both Apple (ObjC) and non-Apple platforms?

**Comment by eeckstein** [comment]
@jrose-apple Not this change. It affects only symbol mangling. AFAIK, archives use the type names stored at runtime.
I changed the runtime type mangling a while ago (but only for non-top-level or generic classes).

**Comment by jrose-apple** [comment]
On Linux we do something else and I can't remember what it is and now I'm worried. :-) @jckarter, @dgrove-oss, @phausler? Is this something we'll have to worry about?

**Comment by phausler** [comment]
This breaks darwin archives but only in a specific case they encode the NSStringFromClass which emits the mangled name in the cases of sub types IIRC.

```swift
import Foundation

class Foo { class Bar { } }

// emulate what NSKeyedArchiver does to store the class name
print(NSStringFromClass(Foo.self))
print(NSStringFromClass(Foo.Bar.self))
```

**Comment by phausler** [comment]
do we have a work-around to support archives that use the old names?

**Comment by eeckstein** [comment]
@phausler You are right, only the names of classes changed if the class is
*) not-top level (as in your example) or
*) generic or
*) private

> do we have a work-around to support archives that use the old names?

No

**Comment by phausler** [comment]
So this will result in external developers having applications that will no longer be able to open archives from previous versions of swift. That is a data loss case. Marking a class private is totally reasonable for developers to do. Changing the type such that it is serialized to disk as something different in the archive will cause users to not be able to open documents they previously saved in older versions of apps.

**Comment by eeckstein** [comment]
Unfortunately in this respect, the archive format is ABI dependent and as we didn't freeze the ABI yet, there will be a breakage.
What we could do is to implement some kind of name translation as a workaround. This would be not so difficult for non-top level and private classes, and a bit of effort for generic classes.
This should be done where the class name lookup is done when reading an archive. Although I don't know where this is.

**Comment by eeckstein** [comment]
As a side note, the name for private classes depend on the module name and the source file name where the class is defined. So if the developer changes any of those, it will also break the archive.

**Comment by jrose-apple** [comment]
Right, we've talked about this before. Private classes are currently out of scope unless the user specifies a custom objc name or uses the NSKeyedArchiver callbacks to control the name. Generic classes have to use the callbacks because they may not exist yet.

**Comment by eeckstein** [comment]
So the only remaining type of classes is non-top level classes. Can we just require to use the callback for those classes as a workaround?
Alternatively we could implement a old->new name translation function in the swift stdlib (It would be implemented by demangling with the old demangler and creating the new name with the new re-mangler.) This function could be called by the archiver to automatically translate names.
What do you think?

**Comment by jrose-apple** [comment]
I don't think users should *ever* have to think about mangling. I'd rather have the old -> new translation as a fallback path.

**Pull request merged**